lib/whitehall_importer/create_file_attachment_revision.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+module WhitehallImporter
+  class CreateFileAttachmentRevision
+    def self.call(*args)
+      new(*args).call
+    end
+
+    def initialize(whitehall_file_attachment, existing_filenames = [])
+      @whitehall_file_attachment = whitehall_file_attachment
+      @existing_filenames = existing_filenames
+    end
+
+    def call
+      check_whitehall_file_attachment_type
+      check_file_requirements
+      blob_revision = create_blob_revision
+
+      FileAttachment::Revision.create!(
+        blob_revision: blob_revision,
+        file_attachment: FileAttachment.create!,
+        metadata_revision: FileAttachment::MetadataRevision.create!(
+          title: whitehall_file_attachment["title"],
+        ),
+      )
+    end
+
+  private
+
+    attr_reader :whitehall_file_attachment, :existing_filenames
+
+    def decorated_file
+      @decorated_file ||= AttachmentFileDecorator.new(download_file, unique_filename)
+    end
+
+    def unique_filename
+      @unique_filename ||= UniqueFilenameService.call(
+        existing_filenames,
+        File.basename(whitehall_file_attachment["url"]),
+      )
+    end
+
+    def download_file
+      file = URI.parse(whitehall_file_attachment["url"]).open
+      if file.is_a?(StringIO)
+        # files less than 10 KB return StringIO (we have to manually cast to a tempfile)
+        Tempfile.new.tap { |tmp| File.write(tmp.path, file.string) }
+      else
+        file
+      end
+    rescue OpenURI::HTTPError
+      raise WhitehallImporter::AbortImportError, "File attachment does not exist: #{whitehall_file_attachment['url']}"
+    end
+
+    def create_blob_revision
+      FileAttachmentBlobService.call(
+        temp_file_attachment: decorated_file,
+        filename: unique_filename,
+      )
+    end
+
+    def check_file_requirements
+      upload_checker = Requirements::FileAttachmentChecker.new(
+        file: decorated_file, title: whitehall_file_attachment["title"],
+      ).pre_upload_issues
+
+      abort_on_issue(upload_checker.issues)
+    end
+
+    def check_whitehall_file_attachment_type
+      return if whitehall_file_attachment["type"] == "FileAttachment"
+
+      raise WhitehallImporter::AbortImportError, "Unsupported file attachment: #{whitehall_file_attachment['type']}"
+    end
+
+    def abort_on_issue(issues)
+      return if issues.empty?
+
+      raise WhitehallImporter::AbortImportError, issues.first.message(style: "form")
+    end
+  end
+
+  class AttachmentFileDecorator < SimpleDelegator
+    attr_reader :original_filename, :content_type
```

Comment:
might be prudent to explicitly set @content_type to nil to show it's intentional

Reply:
Sure